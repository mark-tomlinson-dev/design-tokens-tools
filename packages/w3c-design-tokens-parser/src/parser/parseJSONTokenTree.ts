import { Result } from '@swan-io/boxed';
import {
  matchIsToken,
  matchIsGroup,
  ALIAS_PATH_SEPARATOR,
  JSONTokenTree,
} from 'design-tokens-format-module';

import { traverseJSONValue } from '../utils/traverseJSONValue.js';
import { parseRawToken } from './token/parseRawToken.js';
import { ValidationError } from '../utils/validationError.js';
import { parseRawGroup } from './group/parseRawGroup.js';
import { AnalyzedToken } from './token/AnalyzedToken.js';
import { AnalyzedGroup } from './group/AnalyzedGroup.js';
import { parseTreeNode } from './tree/parseTreeNode.js';
import { makeUniqueId } from '../utils/uniqueId.js';

export type AnalyzedTokenResult = Result<AnalyzedToken, Array<ValidationError>>;
export type AnalyzedGroupResult = Result<AnalyzedGroup, Array<ValidationError>>;

export function parseJSONTokenTree(root: unknown): Result<
  {
    tokenTree: JSONTokenTree;
    tokens: [Array<AnalyzedToken>, Array<ValidationError>];
    groups: [Array<AnalyzedGroup>, Array<ValidationError>];
  },
  ValidationError[]
> {
  return parseTreeNode(root, {
    varName: '[root]',
    nodeId: '',
    path: [],
  }).flatMap((jsonTokenTree) => {
    const analyzedTokens: Array<AnalyzedTokenResult> = [];
    const analyzedGroups: Array<AnalyzedGroupResult> = [];

    traverseJSONValue(jsonTokenTree, (value, rawPath) => {
      const nodeId = makeUniqueId();

      if (matchIsToken(value)) {
        analyzedTokens.push(
          parseRawToken(value, {
            jsonTokenTree,
            nodeId: nodeId,
            path: rawPath,
            varName: `${rawPath.join(ALIAS_PATH_SEPARATOR)}`,
          }),
        );

        return false;
      } else if (matchIsGroup(value)) {
        // Skip root group
        if (rawPath.length > 0) {
          analyzedGroups.push(
            parseRawGroup(value, {
              nodeId: nodeId,
              path: rawPath,
              varName: `${rawPath.join(ALIAS_PATH_SEPARATOR)}`,
            }),
          );
        }

        return true;
      }
      return false;
    });

    // Gather both valid and errored results
    return Result.Ok({
      tokenTree: jsonTokenTree as JSONTokenTree,
      tokens: analyzedTokens.reduce<
        [Array<AnalyzedToken>, Array<ValidationError>]
      >(
        (acc, res) => {
          res.isOk() ? acc[0].push(res.get()) : acc[1].push(...res.getError());
          return acc;
        },
        [[], []],
      ),
      groups: analyzedGroups.reduce<
        [Array<AnalyzedGroup>, Array<ValidationError>]
      >(
        (acc, res) => {
          res.isOk() ? acc[0].push(res.get()) : acc[1].push(...res.getError());
          return acc;
        },
        [[], []],
      ),
    });
  });
}
