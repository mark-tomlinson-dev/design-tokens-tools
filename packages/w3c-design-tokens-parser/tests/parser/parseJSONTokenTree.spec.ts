import { describe, expect, it } from 'vitest';
import { JSONTokenTree } from 'design-tokens-format-module';

import { parseJSONTokenTree } from '../../src/parser/parseJSONTokenTree';
import {
  borderToken,
  colorToken,
  cubicBezierToken,
  dimensionToken,
  durationToken,
  fontFamilyToken,
  gradientToken,
  numberFontWeightToken,
  numberToken,
  shadowToken,
  stringFontWeightToken,
  strokeStyleToken,
  transitionToken,
  typographyToken,
} from '../_fixtures/tokens';

describe('parseJSONTokenTree', () => {
  it('should parse a token tree of raw values of all types', () => {
    const tree: JSONTokenTree = {
      borderToken,
      colorToken,
      cubicBezierToken,
      dimensionToken,
      durationToken,
      fontFamilyToken,
      stringFontWeightToken,
      numberFontWeightToken,
      gradientToken,
      numberToken,
      shadowToken,
      strokeStyleToken,
      transitionToken,
      typographyToken,
    };

    const analyzedResult = parseJSONTokenTree(tree).match({
      Ok: (result) => result,
      Error: (errors) => {
        throw new Error(errors.map((e) => e.message).join('\n'));
      },
    });

    expect(analyzedResult).toMatchSnapshot();
  });
  it('should parse a token tree with references to other tokens', () => {
    const tree: JSONTokenTree = {
      base: {
        primary: colorToken,
      },
      semantic: {
        primary: {
          $value: '{base.primary}',
        },
      },
    };

    const analyzedResult = parseJSONTokenTree(tree).match({
      Ok: (result) => result,
      Error: (errors) => {
        throw new Error(errors.map((e) => e.message).join('\n'));
      },
    });

    expect(JSON.parse(JSON.stringify(analyzedResult))).toStrictEqual({
      tokenTree: {
        base: {
          primary: {
            $type: 'color',
            $value: '#a82222',
          },
        },
        semantic: {
          primary: {
            $value: '{base.primary}',
          },
        },
      },
      tokens: [
        [
          {
            path: ['base', 'primary'],
            type: 'color',
            value: {
              raw: '#a82222',
              toReferences: [],
            },
          },
          {
            path: ['semantic', 'primary'],
            type: 'color',
            value: {
              raw: '{base.primary}',
              toReferences: [
                {
                  fromTreePath: ['semantic', 'primary'],
                  fromValuePath: [],
                  toTreePath: ['base', 'primary'],
                },
              ],
            },
          },
        ],
        [],
      ],
      groups: [
        [
          {
            path: [],
            childrenCount: 2,
          },
          {
            path: ['base'],
            childrenCount: 1,
          },
          {
            path: ['semantic'],
            childrenCount: 1,
          },
        ],
        [],
      ],
    });
  });
  it('should parse a token tree with invalid tokens', () => {
    const tree: JSONTokenTree = {
      base: {
        // @ts-expect-error
        primary: {
          $type: 'color',
          $value: 'rgba(0, 0, 0, 0.5)',
        },
      },
      semantic: {
        primary: {
          // @ts-expect-error
          $type: 'unknown',
          // @ts-expect-error
          $value: 'some value',
        },
      },
      invalidGroup: {
        // @ts-expect-error
        $description: 42,
      },
    };

    const analyzedResult = parseJSONTokenTree(tree).match({
      Ok: (result) => result,
      Error: (errors) => {
        throw new Error(errors.map((e) => e.message).join('\n'));
      },
    });

    expect(JSON.parse(JSON.stringify(analyzedResult))).toStrictEqual({
      tokenTree: {
        base: { primary: { $type: 'color', $value: 'rgba(0, 0, 0, 0.5)' } },
        semantic: { primary: { $type: 'unknown', $value: 'some value' } },
        invalidGroup: { $description: 42 },
      },
      tokens: [
        [],
        [
          {
            type: 'Value',
            isCritical: false,
            treePath: ['base', 'primary'],
            nodeKey: '$value',
            valuePath: [],
            message:
              'base.primary.$value must start with "#" and have a length of 6 or 8. Got: "rgba(0, 0, 0, 0.5)".',
          },
          {
            type: 'Value',
            isCritical: false,
            treePath: ['semantic', 'primary'],
            nodeKey: '$type',
            valuePath: [],
            message:
              'semantic.primary.$type must be a valid type among: "number", "string", "color", "dimension", "duration", "fontFamily", "fontWeight", "cubicBezier", "border", "strokeStyle", "transition", "shadow", "gradient", "typography". Got "unknown".',
          },
        ],
      ],
      groups: [
        [
          { path: [], childrenCount: 3 },
          { path: ['base'], childrenCount: 1 },
          { path: ['semantic'], childrenCount: 1 },
        ],
        [
          {
            type: 'Type',
            isCritical: false,
            treePath: ['invalidGroup'],
            nodeKey: '$description',
            valuePath: [],
            message:
              'invalidGroup.$description must be a string. Got "number".',
          },
        ],
      ],
    });
  });
  it('should parse a token tree containing a circular reference', () => {
    const tree: JSONTokenTree = {
      base: {
        $type: 'color',
        primary: {
          $value: '{base.primary}',
        },
      },
    };

    const analyzedResult = parseJSONTokenTree(tree).match({
      Ok: (result) => result,
      Error: (errors) => {
        throw new Error(errors.map((e) => e.message).join('\n'));
      },
    });

    expect(JSON.parse(JSON.stringify(analyzedResult))).toStrictEqual({
      tokenTree: {
        base: { $type: 'color', primary: { $value: '{base.primary}' } },
      },
      tokens: [
        [],
        [
          {
            isCritical: false,
            message: 'Circular references detected.',
            referenceToTreePath: ['base', 'primary'],
            type: 'Computation',
            treePath: ['base', 'primary'],
            valuePath: [],
          },
        ],
      ],
      groups: [
        [
          { path: [], childrenCount: 1 },
          { path: ['base'], tokenType: 'color', childrenCount: 1 },
        ],
        [],
      ],
    });
  });
});
