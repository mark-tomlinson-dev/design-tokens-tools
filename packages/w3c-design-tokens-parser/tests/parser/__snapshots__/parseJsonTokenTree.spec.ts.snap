// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseJSONTokenTree > should parse a token tree of raw values of all types 1`] = `
{
  "groups": [
    [],
    [],
  ],
  "tokenTree": {
    "borderToken": {
      "$type": "border",
      "$value": {
        "color": "#676767",
        "style": "solid",
        "width": "1px",
      },
    },
    "colorToken": {
      "$type": "color",
      "$value": "#a82222",
    },
    "cubicBezierToken": {
      "$type": "cubicBezier",
      "$value": [
        0,
        1,
        1,
        0,
      ],
    },
    "dimensionToken": {
      "$type": "dimension",
      "$value": "12px",
    },
    "durationToken": {
      "$type": "duration",
      "$value": "1s",
    },
    "fontFamilyToken": {
      "$type": "fontFamily",
      "$value": "Arial",
    },
    "gradientToken": {
      "$type": "gradient",
      "$value": [
        {
          "color": "#000000",
          "position": 0,
        },
        {
          "color": "#ffffff",
          "position": 1,
        },
      ],
    },
    "numberFontWeightToken": {
      "$type": "fontWeight",
      "$value": 700,
    },
    "numberToken": {
      "$type": "number",
      "$value": 12,
    },
    "shadowToken": {
      "$type": "shadow",
      "$value": {
        "blur": "3px",
        "color": "#000000",
        "offsetX": "1px",
        "offsetY": "2px",
        "spread": "4px",
      },
    },
    "stringFontWeightToken": {
      "$type": "fontWeight",
      "$value": "bold",
    },
    "strokeStyleToken": {
      "$type": "strokeStyle",
      "$value": "solid",
    },
    "transitionToken": {
      "$type": "transition",
      "$value": {
        "delay": "0s",
        "duration": "1s",
        "timingFunction": [
          0,
          0.1,
          0.7,
          0.5,
        ],
      },
    },
    "typographyToken": {
      "$type": "typography",
      "$value": {
        "fontFamily": "Arial",
        "fontSize": "12px",
        "fontWeight": "bold",
        "letterSpacing": "1px",
        "lineHeight": 1.5,
      },
    },
  },
  "tokens": [
    [
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "borderToken",
        ],
        "type": "border",
        "value": {
          "raw": {
            "color": "#676767",
            "style": "solid",
            "width": "1px",
          },
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "colorToken",
        ],
        "type": "color",
        "value": {
          "raw": "#a82222",
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "cubicBezierToken",
        ],
        "type": "cubicBezier",
        "value": {
          "raw": [
            0,
            1,
            1,
            0,
          ],
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "dimensionToken",
        ],
        "type": "dimension",
        "value": {
          "raw": "12px",
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "durationToken",
        ],
        "type": "duration",
        "value": {
          "raw": "1s",
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "fontFamilyToken",
        ],
        "type": "fontFamily",
        "value": {
          "raw": "Arial",
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "stringFontWeightToken",
        ],
        "type": "fontWeight",
        "value": {
          "raw": "bold",
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "numberFontWeightToken",
        ],
        "type": "fontWeight",
        "value": {
          "raw": 700,
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "gradientToken",
        ],
        "type": "gradient",
        "value": {
          "raw": [
            {
              "color": "#000000",
              "position": 0,
            },
            {
              "color": "#ffffff",
              "position": 1,
            },
          ],
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "numberToken",
        ],
        "type": "number",
        "value": {
          "raw": 12,
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "shadowToken",
        ],
        "type": "shadow",
        "value": {
          "raw": {
            "blur": "3px",
            "color": "#000000",
            "offsetX": "1px",
            "offsetY": "2px",
            "spread": "4px",
          },
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "strokeStyleToken",
        ],
        "type": "strokeStyle",
        "value": {
          "raw": "solid",
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "transitionToken",
        ],
        "type": "transition",
        "value": {
          "raw": {
            "delay": "0s",
            "duration": "1s",
            "timingFunction": [
              0,
              0.1,
              0.7,
              0.5,
            ],
          },
          "toReferences": [],
        },
      },
      {
        "description": undefined,
        "extensions": undefined,
        "path": [
          "typographyToken",
        ],
        "type": "typography",
        "value": {
          "raw": {
            "fontFamily": "Arial",
            "fontSize": "12px",
            "fontWeight": "bold",
            "letterSpacing": "1px",
            "lineHeight": 1.5,
          },
          "toReferences": [],
        },
      },
    ],
    [],
  ],
}
`;
